awx/api/serializers.py

```diff
@@ -3375,20 +3406,34 @@ def validate(self, attrs):
         JT_job_tags = obj.job_tags
         JT_skip_tags = obj.skip_tags
         JT_inventory = obj.inventory
-        JT_credential = obj.credential
         JT_verbosity = obj.verbosity
-        extra_credentials = attrs.pop('extra_credentials', None)
+        credentials = attrs.pop('credentials', None)
         attrs = super(JobLaunchSerializer, self).validate(attrs)
         obj.extra_vars = JT_extra_vars
         obj.limit = JT_limit
         obj.job_type = JT_job_type
         obj.skip_tags = JT_skip_tags
         obj.job_tags = JT_job_tags
         obj.inventory = JT_inventory
-        obj.credential = JT_credential
         obj.verbosity = JT_verbosity
-        if extra_credentials is not None:
-            attrs['extra_credentials'] = extra_credentials
+        if credentials is not None:
+            attrs['credentials'] = credentials
+
+        # fill passwords dict with request data passwords
+        passwords_needed = []
+        for cred in (attrs['credentials'] or obj.credentials.all()):
```

Comment:
`obj.credentials.exists()`

Reply:
@ryanpetrello actually, I don't know what you're doing here.

```
In [3]: [1, 2] or node.credentials.all()
Out[3]: [1, 2]

In [4]: node.credentials.all()
Out[4]: SELECT "main_credential"."id",
       "main_credential"."created",
       "main_credential"."modified",
       "main_credential"."description",
       "main_credential"."created_by_id",
       "main_credential"."modified_by_id",
       "main_credential"."name",
       "main_credential"."credential_type_id",
       "main_credential"."organization_id",
       "main_credential"."inputs",
       "main_credential"."admin_role_id",
       "main_credential"."use_role_id",
       "main_credential"."read_role_id"
FROM "main_credential"
INNER JOIN "main_workflowjobtemplatenode_credentials" ON ("main_credential"."id" = "main_workflowjobtemplatenode_credentials"."credential_id")
WHERE "main_workflowjobtemplatenode_credentials"."workflowjobtemplatenode_id" = 1
ORDER BY "main_credential"."name" ASC
LIMIT 21

Execution time: 0.001223s [Database: default]

[<Credential: Alan AWS-5>, <Credential: AWS creds-8>, <Credential: bitbucket-1>]
```

I thought you were combining the 2 lists, it looks like you might be assuming that the prompted list blows away the JT list. Could you clarify what your expected behavior is?

Also, this would be a server error because `1.passwords_needed`

Reply:
@AlanCoding I'm pretty sure I've got this right as-is;  I have some tests in `functional/api/test_deprecated_credential_assignment.py` that test it specifically.  At this point in execution `attrs['credentials']` is a list of actual `Credential` model objects (because DRF has converted it for us in `to_internal_value`:

![screen shot 2017-11-10 at 9 55 51 am](https://user-images.githubusercontent.com/214912/32664096-321e6dbc-c5fe-11e7-8fca-8f0222ef7867.png)

So the idea behind this block is to *first* look at the list of prompted credentials, and then fall back to the list of credentials on the JT.

Reply:
@AlanCoding I found a bug in this code, which I'm going to fix:

1.  Make a JT w/ an SSH cred that prompts for a password.
2.  Launch with `{"credentials": []}` (meaning, launch, but with zero credentials).
3.  `HTTP 400` requiring the SSH password (wrong behavior).

Reply:
@ryanpetrello In the view, it is clearly assumed to be a list of primary keys, not a list of objects.

https://github.com/ryanpetrello/awx/blob/47159b083cc32ec72927ed6621ba91a914dde5db/awx/api/views.py#L2799-L2802

If you want my preference, I would love love love for it to start as a list of objects, and remain that way forever and ever.

But it looks like we still haven't fully flushed out type consistency of the credential list as of quite yet.